src/sentry/api/endpoints/chunk.py

```diff
@@ -0,0 +1,284 @@
+
+        if size > MAX_REQUEST_SIZE:
+            return Response({'error': 'Request too large'},
+                            status=status.HTTP_400_BAD_REQUEST)
+
+        if len(files) > MAX_CHUNKS_PER_REQUEST:
+            return Response({'error': 'Too many chunks'},
+                            status=status.HTTP_400_BAD_REQUEST)
+
+        for chunk in files:
+            # Here we create the actual blob
+            blob = FileBlob.from_file(chunk)
+            # Add ownership to the blob here
+            try:
+                with transaction.atomic():
+                    FileBlobOwner.objects.create(
+                        organization=organization,
+                        blob=blob
+                    )
+            except IntegrityError:
+                pass
+            if blob.checksum not in checksum_list:
+                # We do not clean up here since we have a cleanup job
+                return Response({'error': 'Checksum missmatch'},
+                                status=status.HTTP_400_BAD_REQUEST)
+
+        return Response(status=status.HTTP_200_OK)
+
+
+class ChunkAssembleEndpoint(OrganizationEndpoint):
+    permission_classes = (OrganizationReleasePermission, )
+
+    def _create_file_response(self, state, missing_chunks=[]):
+        """
+        Helper function to create response for assemble endpoint
+        """
+        return {
+            'state': state,
+            'missingChunks': missing_chunks
+        }
+
+    def _check_chunk_ownership(self, organization, file_blobs, chunks, file_exists):
+        # Check the ownership of these blobs with the org
+        all_owned_blobs = FileBlobOwner.objects.filter(
+            blob__in=file_blobs,
+            organization=organization
+        ).select_related('blob').all()
+
+        owned_blobs = []
+        for owned_blob in all_owned_blobs:
+            owned_blobs.append((owned_blob.blob.id, owned_blob.blob.checksum))
+
+        # If the request does not cotain any chunks for a file
+        # we return nothing since this should never happen only
+        # if the client sends an invalid request
+        if len(chunks) == 0:
+            return self._create_file_response(
+                ChunkFileState.NOT_FOUND
+            )
+        # Only if this org already has the ownership of all blobs
+        # and the count of chunks is the same as in the request
+        # and the file already exists, we say this file is OK
+        elif len(file_blobs) == len(owned_blobs) == len(chunks) and file_exists:
+            return self._create_file_response(
+                ChunkFileState.OK
+            )
+        # If the length of owned and sent chunks is not the same
+        # we return all missing blobs
+        elif len(owned_blobs) != len(chunks):
+            # Create a missing chunks array which we return as response
+            # so the client knows which chunks to reupload
+            missing_chunks = list(chunks)
```

Comment:
make this into `missing_chunks = set(chunks)` and then remove items with `missing_chunks.discard(blob[1])`. Faster and easier (`O(1)` vs `O(n something)`)